### PR TITLE
Add user plugin

### PR DESCRIPTION
**Pull request by ianwremmel**

### CODE REVIEW

File: packages/plugin-user/src/user-uuid-store.js [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    const p2 = this.getByEmail(user.emailAddress)
+      .then((u) => usersByEmail.get(this).set(user.emailAddress, Object.assign({}, u, user)))
+      .catch(() => usersByEmail.get(this).set(user.emailAddress, Object.assign({}, user)));
+
+    return Promise.all([p1, p2]);
+  }
+
+  /**
+   * Retrievves the specified user object from the store
+   * @param {string} id
```
Comment (Josh-Dykstra) [inline packages/plugin-user/src/user-uuid-store.js]: or an email :)

File: packages/plugin-user/src/user-uuid-store.js [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+/**
+ * @class
+ */
+export default class UserUUIDStore {
+  /**
+   * @constructs {UserUUIDStore}
+   */
+  constructor() {
+    usersByEmail.set(this, new Map());
+    usersById.set(this, new Map());
```
Comment (Josh-Dykstra) [inline packages/plugin-user/src/user-uuid-store.js]: I've seen you use this pattern before. Was it to ensure there was one instance?

File: packages/plugin-user/src/user.js [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+    })
+      .then((res) => res.body);
+  },
+
+  _extractUUID: function _extractUUID(user) {
+    return user.entryUUID || user.id || user;
+  },
+
+  _extractEmailAddress: function _extractEmailAddress(user) {
+    return user.email || user.emailAddress || user.entryEmail || user;
```
Comment (Josh-Dykstra) [inline packages/plugin-user/src/user.js]: Any other potential things that can be thrown at us? :)

File: packages/plugin-user/test/integration/spec/user.js [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    let email;
+    beforeEach(() => {
+      email = makeEmailAddress();
+    });
+
+    it(`maps an email address to a uuid`, () => assert.eventually.equal(spark.user.asUUID(user2, {force: true}), user2.id));
+
+    it(`maps an email address for a non-existent user to a fake uuid`, () => assert.eventually.match(spark.user.asUUID(email), patterns.uuid)
```
Comment (Josh-Dykstra) [inline packages/plugin-user/test/integration/spec/user.js]: Seems like this test doesn't return a fake uuid and instead sets userExists to false

File: packages/plugin-user/src/user-uuid-store.js (hunk 2 above)
Comment (ianwremmel) [inline packages/plugin-user/src/user-uuid-store.js]: Not exactly. It's a way to ensure private data for the class. the only way to access data in those maps is to call the corresponding methods on an instance of UserUUIDStore.

Basically, WeakMaps let us associate arbitrary data with an instance without modifying that instance (and therefor deoptimizing it).

File: packages/plugin-user/test/integration/spec/user.js (hunk 4 above)
Comment (Josh-Dykstra) [inline packages/plugin-user/test/integration/spec/user.js]: I see now we are checking that there is an added entry in the store that has the userExists=false flag

File: packages/plugin-user/src/user.js [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+  get() {
+    return this.request({
+      service: `conversation`,
+      resource: `users`
+    })
+      .then((res) => res.body)
+      .then(tap((user) => this.recordUUID(user)));
+  },
+
+  @oneFlight((email, options) => email + String(options.create))
```
Comment (Josh-Dykstra) [inline packages/plugin-user/src/user.js]: This is so cool.

File: packages/plugin-user/src/user-uuid-store.js (hunk 2 above)
Comment (Josh-Dykstra) [inline packages/plugin-user/src/user-uuid-store.js]: Ok that's right. Thanks.

**Comment by Josh-Dykstra** [comment]
LGTM

**Pull request merged**